### PR TITLE
Enhance CURLOPT_FTP_SKIP_PASV_IP

### PR DESCRIPTION
**Pull request by tenzap**
This is a refresh of https://github.com/curl/curl/pull/1470
I took the original authors PR, squashed it, and tried to refresh it towards the latest version of curl

URL_FTP_SKIP_PASV_IP_ALWAYS instructs libcurl to ignore the IP address the server suggests in its 227-response to libcurl's PASV command when libcurl connects the data connection. Instead libcurl will re-use the same IP address it already uses for the control connection. But it will use the port number from the 227-response.

The allowed values are:

- CURL_FTP_SKIP_PASV_IP_NEVER Always use the IP returned by the server.

- CURL_FTP_SKIP_PASV_IP_ALWAYS Always ignore the IP returned by the server.

- CURL_FTP_SKIP_PASV_IP_IF_NOT_ROUTABLE Only ignore the returned IP if it is within the private address space (see RFC1918), which is a common server configuration error.

Closes: #1455

### CODE REVIEW

**Comment by tenzap** [comment]
@bagder maybe ?

File: src/tool_getparam.c [hunk 1]
```diff
@@ -1198,7 +1198,7 @@ ParameterError getparameter(const char *flag, /* f or -long-flag */
         config->ignorecl = toggle;
         break;
       case 'q': /* --ftp-skip-pasv-ip */
-        config->ftp_skip_ip = toggle;
+        config->ftp_pasvp_ip_rule = CURL_FTP_SKIP_PASV_IP_ALWAYS;
```
Comment (bagder) [inline src/tool_getparam.c:1201]: this breaks behavior for users who do `--no-ftp-skip-pasv-ip` to switch off the option

File: src/tool_cfgable.h [hunk 2]
```diff
@@ -177,7 +177,8 @@ struct OperationConfig {
   bool doh_verifystatus;
   bool create_dirs;
   bool ftp_create_dirs;
-  bool ftp_skip_ip;
+  long ftp_pasvp_ip_rule;   /* how to handle the IP address the FTP server
+                               passes on to us */
```
Comment (bagder) [inline src/tool_cfgable.h:181]: I think you should keep treating this as a boolean cmdline option

File: docs/libcurl/symbols-in-versions [hunk 3]
```diff
@@ -38,6 +38,9 @@ CURL_FORMADD_NULL               7.9.8         7.56.0
 CURL_FORMADD_OK                 7.9.8         7.56.0
 CURL_FORMADD_OPTION_TWICE       7.9.8         7.56.0
 CURL_FORMADD_UNKNOWN_OPTION     7.9.8         7.56.0
+CURL_FTP_SKIP_PASV_IP_ALWAYS          7.86.0
+CURL_FTP_SKIP_PASV_IP_IF_NOT_ROUTABLE 7.86.0
+CURL_FTP_SKIP_PASV_IP_NEVER           7.86.0
```
Comment (bagder) [inline docs/libcurl/symbols-in-versions:43]: This cannot be merged until 7.87.0

File: docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3 [hunk 4]
```diff
@@ -45,7 +52,7 @@ abuse by malicious servers.
 
 This option has no effect if PORT, EPRT or EPSV is used instead of PASV.
 .SH DEFAULT
-1 since 7.74.0, was 0 before then.
+CURL_FTP_SKIP_PASV_IP_ALWAYS
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3:55]: This removes the information about the historic change and I think it should not do that.

File: docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3 [hunk 5]
```diff
@@ -29,14 +29,21 @@ CURLOPT_FTP_SKIP_PASV_IP \- ignore the IP address in the PASV response
...
+the data connection. Instead libcurl will re-use the same IP address it already uses
+for the control connection. But it will use the port number from the 227-response.
+The allowed values are:
+.IP CURL_FTP_SKIP_PASV_IP_NEVER
+Always use the IP returned by the server.
+.IP CURL_FTP_SKIP_PASV_IP_ALWAYS
+Always ignore the IP returned by the server.
+.IP CURL_FTP_SKIP_PASV_IP_IF_NOT_ROUTABLE
+Only ignore the returned IP if it is within the private address space (see RFC1918),
+which is a common server configuration error.
```
Comment (bagder) [inline docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3:46]: I think this should also mention from which version this option works.

File: include/curl/curl.h [hunk 6]
```diff
@@ -2202,6 +2202,12 @@ typedef enum {
   /* Convenient "aliases" */
 #define CURLOPT_RTSPHEADER CURLOPT_HTTPHEADER
 
+  /*parameters for the CURLOPT_FTP_SKIP_PASV_IP option */
```
Comment (bagder) [inline include/curl/curl.h:2205]: A space after `/*` perhaps?

**Review by bagder — COMMENTED** [review]
I think we need a test case for this as well, verifying that it works

**Comment by tenzap** [comment]
@Zenju now that I squashed and refreshed your commits, would you please act on the comments?

File: src/tool_cfgable.h (hunk 2 above)
Comment (tenzap) [reply src/tool_cfgable.h:181]: original author permits 3 values, so I don't think bool would fit here.

File: docs/libcurl/symbols-in-versions (hunk 3 above)
Comment (tenzap) [reply docs/libcurl/symbols-in-versions]: changed to 7.87

File: docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3 (hunk 4 above)
Comment (tenzap) [reply docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3]: updated

File: docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3 (hunk 5 above)
Comment (tenzap) [reply docs/libcurl/opts/CURLOPT_FTP_SKIP_PASV_IP.3:46]: updated, but I guess @zenju has better knowledge of what he wanted to do with this 
PR

File: include/curl/curl.h (hunk 6 above)
Comment (tenzap) [reply include/curl/curl.h]: done

**Review by bagder — CHANGES_REQUESTED** [review]
this breaks existing behavior

**Comment by bagder** [comment]
This can't be merged and no response in weeks. I would also like a stronger motivation why we want/need this feature. Letting the server decide IP address is a security risk.

**Pull request closed without merging**